modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java

```diff
@@ -2141,12 +2141,102 @@ public int compare(CodegenProperty one, CodegenProperty another) {
         return m;
     }
 
+    /**
+     * Recursively look in Schema sc for the discriminator discPropName
+     * @param sc The Schema that may contain the discriminator
+     * @param discPropName The String that is the discriminator propertyName in the schema
+     */
+    private boolean discriminatorFound(Schema sc, String discPropName) {
+        Schema refSchema = ModelUtils.getReferencedSchema(this.openAPI, sc);
+        if (refSchema.getProperties() != null) {
+            Schema discSchema = (Schema) refSchema.getProperties().get(discPropName);
+            if (discSchema != null && ModelUtils.isStringSchema(discSchema)) {
+                return true;
+            }
+        }
+        if (ModelUtils.isComposedSchema(refSchema)) {
+            ComposedSchema composedSchema = (ComposedSchema) refSchema;
+            if (composedSchema.getAllOf() != null) {
+                // If our discriminator is in one of the allOf schemas break when we find it
+                for  (Schema allOf: composedSchema.getAllOf()) {
+                    if (discriminatorFound(allOf, discPropName)) {
+                        return true;
+                    }
+                }
+            }
+            if (composedSchema.getOneOf() != null && composedSchema.getOneOf().size() != 0) {
+                // All oneOf definitions must contain the discriminator
+                Integer hasDiscriminatorCnt = 0;
+                for  (Schema oneOf: composedSchema.getOneOf()) {
+                    if (discriminatorFound(oneOf, discPropName)) {
+                        hasDiscriminatorCnt++;
+                    }
+                }
+                if (hasDiscriminatorCnt == composedSchema.getOneOf().size()) {
+                    return true;
+                }
+            }
+            if (composedSchema.getAnyOf() != null && composedSchema.getAnyOf().size() != 0) {
+                // All anyOf definitions must contain the discriminator because a min of one must be selected
+                Integer hasDiscriminatorCnt = 0;
+                for  (Schema anyOf: composedSchema.getAnyOf()) {
+                    if (discriminatorFound(anyOf, discPropName)) {
+                        hasDiscriminatorCnt++;
+                    }
+                }
+                if (hasDiscriminatorCnt == composedSchema.getAnyOf().size()) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Process oneOf and anyOf models in a composed schema and adds them into
+     * their parent's discriminator map if the oneOf and anyOf models contain
+     * the required discriminator. If they don't contain the required
+     * discriminator or the discriminator is the wrong type then an error is
+     * thrown
```

Comment:
This isn't correct to throw an error for missing discriminator. Notice use of `MAY` in the spec and not `MUST`:

In OAS 3.0, a response payload MAY be described to be exactly one of any number of types:

```
MyResponseType:
  oneOf:
  - $ref: '#/components/schemas/Cat'
  - $ref: '#/components/schemas/Dog'
  - $ref: '#/components/schemas/Lizard'
```

which means the payload MUST, by validation, match exactly one of the schemas described by Cat, Dog, or Lizard. In this case, a discriminator _MAY_ act as a "hint" to shortcut validation and selection of the matching schema which may be a costly operation, depending on the complexity of the schema.

Reply:
SO I think you misunderstand when this code is invoked.
If there is no discriminator in the composed schema, this function is not run.
Per here: https://github.com/spacether/openapi-generator/blob/issue_4904_fix_composed_schema_discriminator_map/modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java#L2234
we return early if there is no discriminator.
If we do have a discriminator then check for discriminator presence because it is required epr the spec here: https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#discriminatorObject
And we invoke our code path here: https://github.com/spacether/openapi-generator/blob/issue_4904_fix_composed_schema_discriminator_map/modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java#L2255

Reply:
Please let me know if I can mark this as resolved.

Reply:
I think a comment that this path is intended to be taken only when a schema had a discriminator would be helpful.

Reply:
A comment has been added to the javadoc for this function `This function is only used for composed schemas which have a discriminator`